File: mobile/lib/models/announcement.dart
class Announcement {
  final int id;
  final String title;
  final String content;
  final String priority; // 'normal', 'important', 'urgent'
  final String targetType; // 'all', 'class', 'santri'
  final List<int>? targetIds;
  final bool pushNotification;
  final String? createdBy;
  final DateTime createdAt;
  final bool isRead;

  Announcement({
    required this.id,
    required this.title,
    required this.content,
    required this.priority,
    required this.targetType,
    this.targetIds,
    required this.pushNotification,
    this.createdBy,
    required this.createdAt,
    required this.isRead,
  });

  factory Announcement.fromJson(Map<String, dynamic> json) {
    return Announcement(
      id: json['id'] as int,
      title: json['title'] as String,
      content: json['content'] as String,
      priority: json['priority'] as String,
      targetType: json['target_type'] as String,
      targetIds: json['target_ids'] != null
          ? List<int>.from(json['target_ids'] as List)
          : null,
      pushNotification: json['push_notification'] as bool? ?? false,
      createdBy: json['created_by'] as String?,
      createdAt: DateTime.parse(json['created_at'] as String),
      isRead: json['is_read'] as bool? ?? false,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'title': title,
      'content': content,
      'priority': priority,
      'target_type': targetType,
      'target_ids': targetIds,
      'push_notification': pushNotification,
      'created_by': createdBy,
      'created_at': createdAt.toIso8601String(),
      'is_read': isRead,
    };
  }

  String get priorityLabel {
    switch (priority) {
      case 'urgent':
        return 'Mendesak';
      case 'important':
        return 'Penting';
      case 'normal':
      default:
        return 'Normal';
    }
  }

  String get priorityIcon {
    switch (priority) {
      case 'urgent':
        return '🔴';
      case 'important':
        return '🟠';
      case 'normal':
      default:
        return '🔵';
    }
  }

  Announcement copyWith({
    int? id,
    String? title,
    String? content,
    String? priority,
    String? targetType,
    List<int>? targetIds,
    bool? pushNotification,
    String? createdBy,
    DateTime? createdAt,
    bool? isRead,
  }) {
    return Announcement(
      id: id ?? this.id,
      title: title ?? this.title,
      content: content ?? this.content,
      priority: priority ?? this.priority,
      targetType: targetType ?? this.targetType,
      targetIds: targetIds ?? this.targetIds,
      pushNotification: pushNotification ?? this.pushNotification,
      createdBy: createdBy ?? this.createdBy,
      createdAt: createdAt ?? this.createdAt,
      isRead: isRead ?? this.isRead,
    );
  }
}

File: Backend/app/Models/Announcement.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Factories\HasFactory;

class Announcement extends Model
{
    use HasFactory;

    protected $fillable = [
        'title',
        'content',
        'priority',
        'target_type',
        'target_ids',
        'push_notification',
        'created_by',
    ];

    protected $casts = [
        'target_ids' => 'array',
        'push_notification' => 'boolean',
        'created_at' => 'datetime',
        'updated_at' => 'datetime',
    ];

    /**
     * Relasi ke User yang membuat pengumuman
     */
    public function creator()
    {
        return $this->belongsTo(User::class, 'created_by');
    }

    /**
     * Relasi ke AnnouncementRead (many to many through pivot)
     */
    public function reads()
    {
        return $this->hasMany(AnnouncementRead::class);
    }

    /**
     * Relasi ke Users yang sudah membaca
     */
    public function readByUsers()
    {
        return $this->belongsToMany(User::class, 'announcement_reads')
            ->withTimestamps()
            ->withPivot('read_at');
    }

    /**
     * Scope untuk filter announcements yang relevan untuk user tertentu
     */
    public function scopeForUser($query, $user)
    {
        return $query->where(function ($q) use ($user) {
            $q->where('target_type', 'all')
                ->orWhere(function ($q) use ($user) {
                    // Jika target adalah kelas, cek apakah user punya santri di kelas tersebut
                    $q->where('target_type', 'class')
                        ->whereJsonContains('target_ids', function ($query) use ($user) {
                            // TODO: implementasi relasi user -> santri -> kelas
                        });
                })
                ->orWhere(function ($q) use ($user) {
                    // Jika target adalah santri tertentu, cek apakah user adalah wali santri tersebut
                    $q->where('target_type', 'santri')
                        ->whereJsonContains('target_ids', function ($query) use ($user) {
                            // TODO: implementasi relasi user -> santri
                        });
                });
        });
    }

    /**
     * Check apakah user sudah membaca announcement ini
     */
    public function isReadBy($userId)
    {
        return $this->reads()->where('user_id', $userId)->exists();
    }
}
